File: Trees_Graphs/go/check_balanced.go
package main
import (
  "fmt"
  "types"
  "math"
)


func isBalanced(lchild *types.TreeNode, rchild *types.TreeNode) bool {
  var left, right bool
  if math.Abs(float64(lchild.Value - rchild.Value)) > 1 {
    return false
  }

  if lchild.LeftChild == nil || lchild.RightChild == nil{
    left = true
  } else {
    left = isBalanced(lchild.LeftChild, lchild.RightChild)
  }

  if rchild.LeftChild == nil || rchild.RightChild == nil{
    right = true
  } else {
    right = isBalanced(rchild.LeftChild, rchild.RightChild)
  }

  return left && right
}


func main() {
  array := []int{2, 5, 3, 4, 8, 6, 4, 7}
  root := types.CreateTree(array)
  types.PrintTreeNodes(root)

  fmt.Println(isBalanced(root.LeftChild, root.RightChild))
}

File: Trees_Graphs/python/check_balanced.py
from Tree import *


def get_height(node):
    if node.left_child is None and node.right_child is None:
        return True, 1

    if node.left_child is None:
        left = 1
    else:
        balanced_left, left = get_height(node.left_child)

    if node.right_child is None:
        right = 1
    else:
        balanced_right, right = get_height(node.left_child)

    if balanced_left and balanced_right:
        return True, max(left + 1, right + 1)


def is_balanced(tree):
    root = tree.root

    balanced_left = get_height(root.left_child)[0]
    balanced_right = get_height(root.right_child)[0]

    if not balanced_left or not balanced_right:
        return False

    return True


def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(is_balanced(tree))


if __name__ == '__main__':
    main()
